src/compiler/crystal/semantic/type_guess_visitor.cr

```diff
@@ -566,16 +566,31 @@ module Crystal
         return type if type.is_a?(PointerInstanceType)
       end
 
-      if type = guess_type_call_pointer_malloc_two_args(node)
-        return type
-      end
+      type = guess_type_call_pointer_malloc_two_args(node)
+      return type if type
 
       type = guess_type_call_lib_fun(node)
       return type if type
 
+      # If it's `obj.tap { ... }`, guess from `obj`
+      obj = is_tap_call?(node)
+      if obj
+        return guess_type(obj)
+      end
+
       nil
```

Comment:
No you can't. The oneliner form doesn't filter out nil. I don't remember the details, but asterite explained it in different issues.